File: 03_plotting-library/templates/knee_plot.py
"""
Knee Plot (Ranked Abundance Plot)
==================================
Log-log ranked barcode UMI count plot for single-cell droplet-based scRNA-seq
quality control. Inspired by CellBender's background removal methodology.

The characteristic "knee" shape separates real cells (high UMI) from
background/empty droplets (low UMI). The inflection point marks the
transition between the two populations.

适用数据类型: distribution / ranked_abundance
必需列: 1-D array of UMI counts (sorted descending)
可选参数: inflection_rank, threshold

参考: Fleming et al., Nature Methods 2023 (CellBender);
      Zheng et al., Nature Communications 2017 (Cell Ranger)
"""

import sys
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
from matplotlib.axes import Axes
from pathlib import Path
from typing import Optional, Tuple

sys.path.insert(0, str(Path(__file__).parent))
from base_plot import (
    load_sci_style, save_fig, apply_gallery_polish,
    SEMANTIC_COLORS,
)

# ============ 参数配置 ============
LINE_COLOR = SEMANTIC_COLORS.get("proposed", "#3C5488")  # deep blue
INFLECTION_COLOR = "#E53935"   # red
THRESHOLD_COLOR = "#8491B4"    # muted grey-blue
LINE_WIDTH = 1.8
CELL_PROB_COLOR = SEMANTIC_COLORS.get("positive", "#2E9E44")  # green for prob curve


# ============ Mock data ============

def generate_mock_data(n_cells=2500, n_background=7500, seed=42):
    """Generate realistic ranked barcode UMI counts.

    Simulates a 10x Chromium-style experiment:
    - Real cells: power-law distributed high UMI counts (1000–50000)
    - Background/empty droplets: low UMI counts (1–100)

    Returns
    -------
    ranked_counts : np.ndarray
        UMI counts sorted in descending order (length n_cells + n_background).
    """
    rng = np.random.default_rng(seed)

    # Real cells — power-law / log-normal distribution
    cell_umis = rng.lognormal(mean=np.log(8000), sigma=0.7, size=n_cells)
    cell_umis = np.clip(cell_umis, 500, 60000).astype(int)

    # Background droplets — low counts
    bg_umis = rng.lognormal(mean=np.log(15), sigma=1.0, size=n_background)
    bg_umis = np.clip(bg_umis, 1, 200).astype(int)

    # Combine and sort descending
    all_umis = np.concatenate([cell_umis, bg_umis])
    ranked_counts = np.sort(all_umis)[::-1]

    return ranked_counts


def detect_inflection(ranked_counts):
    """Detect the knee/inflection point using maximum curvature.

    Works in log-log space on the cumulative fraction curve.

    Returns
    -------
    inflection_rank : int
        The 1-indexed rank at the inflection point.
    """
    log_ranks = np.log10(np.arange(1, len(ranked_counts) + 1))
    log_counts = np.log10(np.maximum(ranked_counts, 1))

    # Smooth to reduce noise
    window = max(len(ranked_counts) // 200, 5)
    kernel = np.ones(window) / window
    smoothed = np.convolve(log_counts, kernel, mode="same")

    # Second derivative (curvature proxy)
    d2 = np.diff(smoothed, n=2)
    # Skip edges
    margin = window
    inflection_idx = np.argmin(d2[margin:-margin]) + margin

    return int(10 ** log_ranks[inflection_idx])


def compute_cell_probability(ranked_counts, inflection_rank, steepness=5.0):
    """Compute probability of each barcode being a real cell.

    Uses a sigmoid centered at the inflection point (in log-rank space).

    Returns
    -------
    prob : np.ndarray
        Probability values in [0, 1], same length as ranked_counts.
    """
    log_ranks = np.log10(np.arange(1, len(ranked_counts) + 1))
    log_inflection = np.log10(inflection_rank)
    # Sigmoid: high prob for rank < inflection, low for rank > inflection
    prob = 1.0 / (1.0 + np.exp(steepness * (log_ranks - log_inflection)))
    return prob


# ============ Plotting ============

def plot(
    ranked_counts,
    inflection_rank=None,
    threshold=None,
    show_probability=True,
    xlabel="Barcode rank",
    ylabel="UMI counts",
    title=None,
    figsize=None,
    save_path=None,
    ax=None,
    preset="publication",
):
    """
    Draw a log-log ranked abundance (knee) plot.

    Parameters
    ----------
    ranked_counts : array-like
        UMI counts per barcode, sorted in descending order.
    inflection_rank : int or None
        Rank at the knee point. If None, auto-detected.
    threshold : float or None
        Minimum UMI cutoff line to draw.
    show_probability : bool
        Whether to overlay the cell-probability sigmoid on a secondary y-axis.
    xlabel : str
        X-axis label.
    ylabel : str
        Y-axis label.
    title : str or None
        Plot title.
    figsize : tuple or None
        Figure size in inches.
    save_path : str or None
        If given, save figure to this path.
    ax : matplotlib Axes or None
        Existing axes to draw on. If None, a new figure is created.
    preset : str
        Style preset: 'publication' | 'gallery' | 'presentation' | 'draft'

    Returns
    -------
    ax : matplotlib Axes
    """
    load_sci_style(preset)

    ranked_counts = np.asarray(ranked_counts, dtype=float)
    n = len(ranked_counts)
    ranks = np.arange(1, n + 1)

    # Auto-detect inflection
    if inflection_rank is None:
        inflection_rank = detect_inflection(ranked_counts)

    # --- Create figure ---
    external_ax = ax is not None
    if ax is None:
        if figsize is None:
            figsize = (8, 5.6) if preset == "gallery" else (5, 4)
        fig, ax = plt.subplots(figsize=figsize)
    else:
        fig = ax.figure

    # --- Main ranked curve (log-log) ---
    ax.plot(
        ranks, ranked_counts,
        color=LINE_COLOR, linewidth=LINE_WIDTH,
        solid_capstyle="round", zorder=3,
    )
    ax.set_xscale("log")
    ax.set_yscale("log")

    # --- Inflection point marker ---
    inflection_count = ranked_counts[min(inflection_rank - 1, n - 1)]
    ax.axvline(
        x=inflection_rank, color=INFLECTION_COLOR,
        linestyle="--", linewidth=1.2, alpha=0.85, zorder=4,
    )
    ax.annotate(
        f"Knee: rank {inflection_rank:,}\n({int(inflection_count):,} UMIs)",
        xy=(inflection_rank, inflection_count),
        xytext=(inflection_rank * 3, inflection_count * 0.3),
        fontsize=plt.rcParams.get("font.size", 7) - 1,
        color=INFLECTION_COLOR,
        arrowprops=dict(
            arrowstyle="->", color=INFLECTION_COLOR,
            lw=0.8, connectionstyle="arc3,rad=0.2",
        ),
        bbox=dict(boxstyle="round,pad=0.3", fc="white", ec=INFLECTION_COLOR,
                  alpha=0.9, lw=0.5),
        zorder=5,
    )

    # --- Threshold line ---
    if threshold is not None:
        ax.axhline(
            y=threshold, color=THRESHOLD_COLOR,
            linestyle=":", linewidth=1.0, alpha=0.7, zorder=2,
        )
        ax.text(
            n * 0.6, threshold * 1.15,
            f"Min UMI = {int(threshold):,}",
            fontsize=plt.rcParams.get("font.size", 7) - 2,
            color=THRESHOLD_COLOR, ha="center",
        )

    # --- Cell probability on secondary axis ---
    if show_probability:
        ax2 = ax.twinx()
        prob = compute_cell_probability(ranked_counts, inflection_rank)
        ax2.plot(
            ranks, prob,
            color=CELL_PROB_COLOR, linewidth=1.2, alpha=0.65,
            linestyle="-.", zorder=2,
        )
        ax2.set_ylabel("P(cell)", color=CELL_PROB_COLOR)
        ax2.set_ylim(-0.05, 1.05)
        ax2.set_xscale("log")
        ax2.tick_params(axis="y", labelcolor=CELL_PROB_COLOR)
        for spine in ax2.spines.values():
            spine.set_color(CELL_PROB_COLOR)
            spine.set_linewidth(0.6)

    # --- Axes labels ---
    ax.set_xlabel(xlabel)
    ax.set_ylabel(ylabel)
    if title:
        ax.set_title(title, fontweight="bold", pad=8)

    # --- Polish ---
    apply_gallery_polish(ax)

    # --- Save ---
    if save_path:
        save_fig(ax.figure if external_ax else fig,
                 Path(save_path).stem, transparent=False)

    return ax


# ============ CLI / Demo ============

if __name__ == "__main__":
    from base_plot import load_sci_style, save_fig

    load_sci_style("gallery")

    ranked = generate_mock_data(n_cells=2500, n_background=7500, seed=42)
    knee = detect_inflection(ranked)

    fig, ax_main = plt.subplots(figsize=(8, 5.6))
    ax_main = plot(
        ranked,
        inflection_rank=knee,
        threshold=500,
        show_probability=True,
        xlabel="Barcode rank",
        ylabel="UMI counts",
        title="Knee Plot — Droplet scRNA-seq Barcode Filtering",
        preset="gallery",
        ax=ax_main,
    )

    name = Path(__file__).stem
    save_fig(ax_main.figure, name, dpi=180, fmt="both")
    plt.close(ax_main.figure)

    print(f"Knee plot saved. Inflection rank: {knee:,}")

File: 03_plotting-library/templates/knee_plot.R
#' Knee Plot (Ranked Abundance with Inflection)
#'
#' Log-log ranked abundance plot for cell barcode / UMI filtering.
#' Detects the inflection ("knee") point separating real cells from
#' background. Inspired by CellBender (Fleming et al., Nature Methods 2023)
#' and EmptyDrops (Lun et al., PLoS Comput Biol 2019).
#'
#' Required columns: barcode_id, umi_count
#' Optional columns: is_cell (logical), cell_prob (0-1)

suppressPackageStartupMessages({
  library(ggplot2)
})

tryCatch(source("/app/r_libs/style/base_plot.R"),
         error = function(e) tryCatch(source("/app/r_libs/style/base_plot.R"),
         error = function(e) source("base_plot.R")))

find_knee <- function(log_counts, log_rank) {
  # Max curvature in log-log space
  n <- length(log_counts)
  if (n < 10) return(n)
  dx <- diff(log_rank)
  dy <- diff(log_counts)
  ds <- sqrt(dx^2 + dy^2)
  tx <- cumsum(dx / ds)
  ty <- cumsum(dy / ds)
  ddx <- diff(tx)
  ddy <- diff(ty)
  dds <- sqrt(ddx^2 + ddy^2)
  kappa <- abs(ddx * ty[-1] - ddy * tx[-1]) / (dds^3)
  kappa[is.na(kappa) | is.infinite(kappa)] <- 0
  which.max(kappa) + 1
}

generate_mock_data <- function(n_cells = 2500, n_empty = 7500, seed = 42) {
  set.seed(seed)

  # Real cells: higher UMI counts
  cell_counts <- rlnorm(n_cells, meanlog = 7, sdlog = 1.2)
  # Background: lower counts
  empty_counts <- rlnorm(n_empty, meanlog = 3, sdlog = 0.8)

  all_counts <- c(cell_counts, empty_counts)
  all_counts <- sort(all_counts, decreasing = TRUE)

  data.frame(
    barcode_id = seq_along(all_counts),
    umi_count = round(all_counts),
    is_cell = c(rep(TRUE, n_cells), rep(FALSE, n_empty))[order(c(
      seq_len(n_cells), n_cells + seq_len(n_empty)
    ), decreasing = TRUE)],
    cell_prob = c(
      runif(n_cells, 0.7, 1.0),
      runif(n_empty, 0, 0.3)
    ),
    stringsAsFactors = FALSE
  )
}

plot_knee <- function(df,
                      rank = "barcode_id",
                      count = "umi_count",
                      is_cell = "is_cell",
                      cell_prob = NULL,
                      show_inflection = TRUE,
                      base_size = 14) {

  df[[rank]] <- as.numeric(df[[rank]])
  df <- df[order(df[[count]], decreasing = TRUE), ]
  df[[rank]] <- seq_len(nrow(df))

  # Build aes conditionally (ggplot2 3.5+ forbids if() inside aes)
  point_aes <- aes(x = .data[[rank]], y = .data[[count]])
  has_is_cell <- !is.null(is_cell) && is_cell %in% names(df) && is.character(is_cell) && length(is_cell) == 1
  if (has_is_cell) {
    point_aes <- aes(x = .data[[rank]], y = .data[[count]], color = .data[[is_cell]])
  }

  p <- ggplot(df, point_aes) +
    geom_point(size = 0.5, alpha = 0.6, shape = 16) +
    scale_x_log10(labels = scales::comma) +
    scale_y_log10(labels = scales::comma) +
    annotation_logticks(sides = "bl", linewidth = 0.2)

  if (has_is_cell) {
    p <- p + scale_color_manual(values = c("TRUE" = "#E64B35", "FALSE" = "#CFD5E2"),
                                 name = "Cell", labels = c("Empty", "Cell"))
  } +
    theme_sci(base_size = base_size) +
    theme(
      legend.position = "top",
      legend.key.height = unit(6, "pt")
    ) +
    labs(x = "Barcode Rank", y = "UMI Count")

  # Mark inflection point
  if (show_inflection) {
    log_c <- log10(df[[count]])
    log_r <- log10(df[[rank]])
    knee_idx <- find_knee(log_c, log_r)
    if (knee_idx > 0 && knee_idx <= nrow(df)) {
      p <- p + annotate("vline", xintercept = df[[rank]][knee_idx],
                        linetype = "dashed", color = "#3C5488", linewidth = 0.5) +
        annotate("label",
                 x = df[[rank]][knee_idx] * 2,
                 y = df[[count]][1] * 0.3,
                 label = paste0("Knee (", scales::comma(df[[rank]][knee_idx]), ")"),
                 size = 3, color = "#3C5488", fill = "white", label.size = 0)
    }
  }

  p
}

# Demo ----
if (sys.nframe() == 0 || isTRUE(getOption("run_demo"))) {
  df <- generate_mock_data()
  p <- plot_knee(df)

  out <- template_out_dir()
  save_demo(p, name = "knee_plot_demo", out_dir = out,
            width = 140, height = 100, units = "mm", dpi = 300)
  message("Demo saved to ", file.path(out, "knee_plot_demo.png"))
}
